### PR TITLE
Use func_wrap in wasmi::Linker

### LINKED ISSUE #3784: Use `wasmi::func_wrap`
Use `func_wrap` instead of `define` and `Func::wrap`. 

This might allow us to refactor, so that we don't have to create a new Linker for every instance.

### PR DESCRIPTION
**Pull request by mariaschett**
Fixes #3784

### CODE REVIEW

File: oak_functions/wasm/src/lib.rs [hunk 1]
```diff
@@ -271,7 +196,91 @@ where
...
+                    from_oak_status(result)
+                },
+            )
+            .expect("failed to define invoke in linker");
+        OakLinker { linker }
+    }
+}
+
 /// Reads the buffer starting at address `buf_ptr` with length `buf_len` from the Wasm memory.
+
```
Comment (conradgrobler) [inline oak_functions/wasm/src/lib.rs:283]: nit: this blank line should probably be removed.

**Review by conradgrobler — APPROVED** [review]
Thanks, looks cleaner now.

File: oak_functions/wasm/src/lib.rs [hunk 2]
```diff
@@ -271,7 +196,91 @@ where
     }
 }
 
+// Exports the functions from oak_functions_abi/src/lib.rs.
+// To instantiate this linker, use the same store used creating the new linter.
```
Comment (conradgrobler) [inline oak_functions/wasm/src/lib.rs:200]: nit: This should probalby be a doccomment, and perhaps the last part could be "use the same store used when creating the new linter."

Comment (mariaschett) [reply oak_functions/wasm/src/lib.rs]: I refactored to also add instantiate to the OakLinker (had to fight the borrow checker a bit). Could you have a second look? Thank you!

File: oak_functions/wasm/src/lib.rs [hunk 3]
```diff
@@ -136,90 +137,133 @@ where
...
         let mut store = wasmi::Store::new(module.engine(), user_state);
+        let linker = OakLinker::new(module.engine(), &mut store);
+        let (instance, store) = linker.instantiate(store, module)?;
+        // TODO(#3785): Add logging and list exports.
+        let wasm_state = Self { instance, store };
+        Ok(wasm_state)
+    }
 
-        let mut linker: wasmi::Linker<UserState<L>> = wasmi::Linker::new(module.engine());
+    // Invokes the Wasm module by calling main.
```
Comment (conradgrobler) [inline oak_functions/wasm/src/lib.rs:149]: nit: pre-existing, but comments directly above a function, struct, etc should be doccomments, event if it is not public.  AFAIK we also generate documentation for non-public code.

**Review by conradgrobler — APPROVED** [review]

File: oak_functions/wasm/src/lib.rs (hunk 3 above)
Comment (mariaschett) [reply oak_functions/wasm/src/lib.rs:149]: Great, I'll refactor some more in the next PRs and do this then.

**Pull request merged**